File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.cc
#include <iostream>
#include <cmath>
using namespace std;
void hanoi(int n, const int& from, const int& temp, const int& to);
int main() {
	int n;
	cin >> n;
	cout << (int)pow(2, n) - 1 << '\n';
	hanoi(n, 1, 2, 3);
	return 0;
}
void hanoi(int n, const int& from, const int& temp, const int& to) {
	if (n == 1) {
		cout << from << ' ' << to << '\n';
		return;
	}
	hanoi(n - 1, from, to, temp);
	cout << from << ' ' << to << '\n';
	hanoi(n - 1, temp, from, to);
}

File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        sb.append((int)Math.pow(2, n) - 1).append("\n");
        hanoi(sb, n, '1', '2', '3');
        System.out.print(sb);
    }

    public static void hanoi(StringBuilder sb, int n, char from, char temp, char to) {
        if (n == 1) {
            sb.append(from).append(' ').append(to).append('\n');
            return;
        }
        hanoi(sb, n - 1, from, to, temp);
        sb.append(from).append(' ').append(to).append('\n');
        hanoi(sb, n - 1, temp, from, to);
    }
}
